### PR TITLE
Upload to riff-raff

### PR DESCRIPTION
**Pull request by AWare**
[node-riffraff-artefact](https://github.com/guardian/node-riffraff-artefact) has been around a while and has some problems which makes it unsuitable for new projects:

- its spelling is not consistent with [sbt-riffraff-artifact](https://github.com/guardian/sbt-riffraff-artifact) (this is mostly just annoying, but this is why this is in a new home)
- it can only handle two deployment artifact collections per project, where one is cloudformation
- it requires zip to be installed on the build agent, mostly precluding the node docker container 
- it is very dependent on your package.json and folder structure in difficult to reason about ways

this PR represents my work in replacing it and will be updated as it gradually progresses

### CODE REVIEW

File: lib/bundle.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        }
+        if (stats.isDirectory()) {
+            resolve("directory");
+            return;
+        }
+        if (stats.isFile()) {
+            resolve("file");
+            return;
+        }
+        console.warn(`${path} must either be a file or directory`, stats.isSymbolicLink() && `(it's a symlink 🔗)`, stats.isBlockDevice() && `(it's a block device 💾)`, stats.isSocket() && `(it's a socket 🔧)`);
```
Comment (philmcmahon) [inline lib/bundle.js]: nice emoji!

**Comment by AWare** [comment]
I think this is ready to release a preview build of to test out.

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,60 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
+    "target": "ES2017",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+     "lib": ["es2017"],                             /* Specify library files to be included in the compilation. */
+    // "allowJs": true,                       /* Allow javascript files to be compiled. */
```
Comment (philmcmahon) [inline tsconfig.json:7]: I guess these should be tidied at some point?

File: src/upload.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+const uploadCompressed = (manifest: Manifest, action: Action) => {
+  if (!action.compress) {
+    throw new Error("Attempted to compress something when compress was false.");
+  }
+  const stream = new PassThrough();
+  compressToStream(action.path, action.compress, stream).catch(_ =>
+    console.log(_)
+  );
+  // stream.on("data", _ => console.log("data", _));
```
Comment (philmcmahon) [inline src/upload.ts]: delete?

File: src/upload.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    name
+  ]
+    .filter(_ => _ != null)
+    .join("/");
+  return s3
+    .upload({
+      ACL: "bucket-owner-read",
+      Body: stream,
+      // Bucket: "riffraff-artifact",
+      Bucket: "alex-w-test-bucket",
```
Comment (philmcmahon) [inline src/upload.ts]: update?

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+export const deploy = async () => {
+  const conf = await getConfig();
+
+  const projectName = conf.projectName;
+  const vcsURL = conf.vcsURL || "https://gu.com";
+  const actions: Action[] = conf.actions;
+
+  const manifest = generateManifest(projectName, vcsURL);
+  console.log(manifest);
```
Comment (philmcmahon) [inline src/index.ts]: is this still needed? I guess so as it has useful info, maybe it needs some context?

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  // upload each action
+  await Promise.all(
+    actions.map(action => {
+      return uploadAction(manifest, action);
+    })
+  );
+
+  // upload rr.y
```
Comment (philmcmahon) [inline src/index.ts]: I think this comment could be less wordy ;)

File: src/upload.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+export const uploadAction = (manifest: Manifest, action: Action) => {
+  if (action.compress) {
+    return uploadCompressed(manifest, action); // we know this isn't false ;_;
+  }
+  return uploadMany(manifest, action);
+};
+
+const uploadMany = async (manifest: Manifest, action: Action) => {
+  const files = await getAllFiles(action.path);
+  console.log(files);
```
Comment (philmcmahon) [inline src/upload.ts]: again does this need some context?

**Comment by philmcmahon** [comment]
Amaaaazing!!! Thanks Alex what a hero. Let me know if you need help testing or something.

**Review by philmcmahon — APPROVED** [review]

**Pull request merged**